R/check_tbl_spl_n.R

```diff
@@ -0,0 +1,124 @@
+#' Check model output data tbl samples contain the appropriate number of samples
+#' for a given compound idx.
+#' @param tbl a tibble/data.frame of the contents of the file being validated. Column types must **all be character**.
+#' @inherit check_tbl_colnames params
+#' @inherit check_tbl_colnames return
+#' @export
+check_tbl_spl_n <- function(tbl, round_id, file_path, hub_path) {
+  config_tasks <- hubUtils::read_config(hub_path, "tasks")
+
+  if (isFALSE(has_spls_tbl(tbl)) || isFALSE(hubUtils::is_v3_config(config_tasks))) {
+    return(skip_v3_spl_check(file_path))
+  }
+
+  hash_tbl <- spl_hash_tbl(tbl, round_id, config_tasks)
+  n_ranges <- get_round_spl_n_ranges(config_tasks, round_id)
+
+  n_tbl <- dplyr::group_by(hash_tbl, .data$compound_idx) %>%
+    dplyr::summarise(
+      n = dplyr::n_distinct(.data$output_type_id),
+      mt_id = unique(.data$mt_id)
+    ) %>%
+    dplyr::left_join(n_ranges, by = "mt_id") %>%
+    dplyr::mutate(
+      less = .data$n < .data$n_min,
+      more = .data$n > .data$n_max,
+      out_range = .data$less | .data$more
+    ) %>%
+    dplyr::filter(.data$out_range)
+
+  check <- nrow(n_tbl) == 0L
+
+  if (check) {
+    details <- NULL
+    errors <- NULL
+  } else {
+    errors <- n_mismatch_errors(
+      n_tbl, hash_tbl, tbl
+    )
+    details <- n_mismatch_details(n_tbl)
+  }
+
+  capture_check_cnd(
+    check = check,
+    file_path = file_path,
+    msg_subject = "Required samples per compound idx task",
+    msg_attribute = NULL,
+    msg_verbs = c("present.", "not present."),
+    details = details,
+    errors = errors
+  )
+}
+
+get_round_spl_n_ranges <- function(config_tasks, round_id) {
+  round_mt <- hubUtils::get_round_model_tasks(config_tasks, round_id)
+  purrr::imap(
+    round_mt,
+    ~ {
+      output_type_id_params <- purrr::pluck(
+        .x,
+        "output_type",
+        "sample",
+        "output_type_id_params"
+      )
+
+      if (is.null(output_type_id_params)) {
+        return(NULL)
+      } else {
+        tibble::tibble(
+          mt_id = .y,
+          n_min = output_type_id_params$min_samples_per_task,
+          n_max = output_type_id_params$max_samples_per_task
+        )
+      }
+    }
+  ) %>%
+    purrr::list_rbind()
+}
+
+n_mismatch_errors <- function(n_tbl, hash_tbl, tbl) {
+  tbl <- tbl[tbl$output_type == "sample", names(tbl) != "value"]
+  purrr::map(
+    purrr::set_names(n_tbl$compound_idx),
+    ~ {
+      spl_d <- hash_tbl$output_type_id[hash_tbl$compound_idx == .x][1L]
+      compound_idx_tbl <- tbl[
+        tbl$output_type_id == spl_d,
+        setdiff(names(tbl), c("output_type_id", "value", "output_type"))
+      ]
+      row <- n_tbl[n_tbl$compound_idx == .x, ] %>% as.vector()
+      list(
+        compound_idx = .x,
+        n = row$n,
+        min_samples_per_task = row$n_min,
+        min_samples_per_task = row$n_max,
+        compound_idx_tbl = compound_idx_tbl
```

Comment:
To ensure number of samples are correct, we can't just count the number of distinct values across the compound task id set. We need to ensure that they have been correctly allocated to the appropriate sample through the value in output type id. And each sample (i.e. unique `output_type_id` value) should have only a single value in each of the compound task id set columns:

In the file you have been testing, the samples do not appear to be correctly specified, for example, sample no 10 (with output_type_id == 10) contains more than one unique value for `scenario_id` when it should contain only one:
```r
tbl |> filter(output_type_id == 10) |> select( origin_date, scenario_id,  location, target) %>% distinct()
# A tibble: 276 × 4
   origin_date scenario_id  location target  
   <chr>       <chr>        <chr>    <chr>   
 1 2024-04-28  A-2024-03-01 51       inc hosp
 2 2024-04-28  B-2024-03-01 51       inc hosp
 3 2024-04-28  C-2024-03-01 51       inc hosp
 4 2024-04-28  D-2024-03-01 51       inc hosp
 5 2024-04-28  E-2024-03-01 51       inc hosp
 6 2024-04-28  F-2024-03-01 51       inc hosp
 7 2024-04-28  A-2024-03-01 05       inc hosp
 8 2024-04-28  B-2024-03-01 05       inc hosp
 9 2024-04-28  C-2024-03-01 05       inc hosp
10 2024-04-28  D-2024-03-01 05       inc hosp
# ℹ 266 more rows
# ℹ Use `print(n = ...)` to see more rows
```

This is throwing off the `sample_n` validation and should have been picked by `check_tbl_spl_compound_tid()` but sadly this is causing an error.

You can investigate the error by looking at the `error` object (as discussed this could be made clearer). Also having a look at the `sample` subset of the`expand_model_out_val_grid(...,  include_sample_ids = TRUE)` output might be able to help in the meantime to see what samples for different compound idxs should look like (until https://github.com/Infectious-Disease-Modeling-Hubs/hubData/issues/40 is implemented.)

Reply:
>  And each sample (i.e. unique output_type_id value) should have only a single value in each of the compound task id set columns:

I thought we were allowing different level of compound id set. For example, in the documentation (https://hubverse.io/en/latest/user-guide/sample-output-type.html#configuration-of-output-type-id) the compound set C is accepted with all the other compound set. 

So here if I have a team that submit a file with another level of `"compound_task_id_set"` (for example: `"origin_date", "location", "target"]`), it should be accepted and different scenario id value should be expected. I tested it with a file where the compound id set is: `["origin_date"]` and it worked (`JHU_UNC-flepiMoP_subset` in my example repo).

Reply:
Hmmm indeed. However this wasn't explicit in the[ instructions for the validation tests](https://docs.google.com/document/d/1vJT8Uw8fQgoY5sajcokZSfnF7M89eiiiwlRXVkCuj2k/edit#heading=h.qli9sktijlj4). It also makes validating a lot more complicated and would need thought as to how to perform such a loose check. 

At the moment, as the tests are set up this is not allowed and I don;t think it will be simple enough to introduce before the planned release so perhaps opening an issue to include tis and think through how it would even work would be the best way to go forward atm.

Reply:
Also origin date would pass because there is normally only a single value per submission file.

Reply:
If we want a quick test that does not take into account the `output_type_id`, you can use the code I use here to calculate the number of samples per each task id group (without taking into account the `"compound_task_id_set"`) but we loose the link with the compound task id. However it will allow tag that the error is just on the ID of the sample not on the total number of samples in the submission file. If we want to provide that information

Reply:
Honestly, I still don't feel that's right. But I think the problem here is that when files have not passed the validation checks for correctly defined samples, trying to count the samples is pointless. I therefore think the  correct way to proceed would be to set up tests so that they are passed sequentially.